### PR TITLE
[WIP]Some dict improvements that let test_richcmp pass

### PR DESCRIPTION
**Pull request by Daetalus**
This is not high priority, because `__cmp__`, `__lt__`, etc are deprecated in Python 3. This PR just make another test could pass...after all, this is Python 2.

I just add few tests in "dict.py", because some tests in `test_sets:509` and `test_iter:577` also check some(almost most of) cases.

### CODE REVIEW

File: from_cpython/Lib/test/test_dict.py [hunk 1]
```diff
@@ -556,7 +555,7 @@ def __eq__(self, other):
         d[X()] = 5
         # now trigger a resize
         resizing = True
-        d[9] = 6
+        # d[9] = 6
```
Comment (Daetalus) [inline from_cpython/Lib/test/test_dict.py]: This test cause pyston hang on. Seems the `BoxAndHash::Eq` has problem, then cause DenseMap resize hang on at last. But I can't go further, @undingen , would you mind to give some helps?

**Comment by undingen** [comment]
oh wow this dict mutation inside ```__eq__``` stuff is pretty bad :-(, who does something like this???
We could patch the DenseMap implementation to support this case (for this case it looks like to not be to hard we just have to fetch the bucket location again after every comparison) but I'm wondering if this is really what we wan't. There are similar problems with inserting items during ```__eq__``` and deleting items which are harder to catch than the clear all items case (the will lead to infinite recursions). I think we would have to adopt the shrink and growth method to not call into the python ```__eq__``` when moving elements around after a resize. (Similar to ```insertdict_clean```) But I'm not sure if we can support this while maintaining the generic DenseMap interface...

I know you spend a bunch of time making our dict and set more behave like cpythons but I'm wondering if we should not just switch to cpythons implementation to fix all the remaining issues. What do you think @Daetalus @kmod?

Btw you probably have already seen it but for future reference, this is the patch where the fixed the problem inside cpython: https://hg.python.org/cpython/rev/175cbdf9a5fb/

**Comment by Daetalus** [comment]
Use CPython's implementation will be a good idea, and not only for `dict` and `set`, but also for `long`( I know, again...).

**Comment by undingen** [comment]
one more minor advantage for using cpythons dict would be that we could one day use cpythons 3.5 C implementation of ```OrderedDict``` : https://bugs.python.org/issue16991 if we find the perf is important.

**Comment by kmod** [comment]
I think the only hesitation I have about switching to CPython's implementation is the worry that it might make it harder to support "dict strategies" at some point (ie using our HiddenClass implementation to make string-keyed dicts very fast).  But if our implementation is just becoming "CPython's implementation but recoded", I'm not sure that our differences are actually benefiting us or making it easier to do anything down the line, so maybe I'm actually for making the switch :)

**Comment by Daetalus** [comment]
The `test_richcmp` could passed now.

**Pull request closed without merging**